clippy_lints/src/needless_cow.rs

```diff
@@ -0,0 +1,384 @@
+    /// If the `Cow` is passed to a function expecting `Cow`, or is mutably borrowed, this will
+    /// still lint.
+    ///
+    /// ### Example
+    /// ```rust
+    /// # use std::borrow::Cow;
+    ///
+    /// # let my_owned_string = String::new();
+    /// # let something = true;
+    /// # let other_thing = true;
+    ///
+    /// let _ = if something {
+    ///     Cow::Owned(my_owned_string);
+    /// } else if other_thing {
+    ///     Cow::Owned(String::new());
+    /// } else {
+    ///     Cow::Borrowed("");
+    /// };
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// # use std::borrow::Cow;
+    ///
+    /// # let my_owned_string = String::new();
+    /// # let something = true;
+    /// # let other_thing = true;
+    ///
+    /// let _ = if something {
+    ///     my_owned_string;
+    /// } else if other_thing {
+    ///     String::new();
+    /// } else {
+    ///     String::new();
+    /// };
+    /// ```
+    #[clippy::version = "1.73.0"]
+    pub NEEDLESS_COW,
+    nursery,
+    "usage of `Cow` with only empty strings in `Borrow` variant"
+}
+impl_lint_pass!(NeedlessCow => [NEEDLESS_COW]);
+
+pub struct NeedlessCow;
+
+impl<'tcx> LateLintPass<'tcx> for NeedlessCow {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        _: FnKind<'tcx>,
+        _: &'tcx FnDecl<'tcx>,
+        _: &'tcx Body<'tcx>,
+        _: Span,
+        def_id: LocalDefId,
+    ) {
+        // Probably unnecessary nowadays considering dummy MIR is created for these functions.
+        if fn_has_unsatisfiable_preds(cx, def_id.to_def_id()) {
+            return;
+        }
+
+        let mir = cx.tcx.optimized_mir(def_id);
+        let mut cow_locals_to_assignments = FxIndexMap::default();
+
+        for local in mir
+            .local_decls
+            .iter_enumerated()
+            .filter(|(_, decl)| is_type_diagnostic_item(cx, decl.ty, sym::Cow))
+            .map(|(i, _)| i)
+        {
+            cow_locals_to_assignments.insert(local, vec![]);
+        }
+
+        let mut results = NeedlessCowAnalysis {
+            map: Map::from_filter(cx.tcx, mir, |ty| ty.peel_refs().is_str(), None),
+        }
+        .wrap()
+        .into_engine(cx.tcx, mir)
+        .iterate_to_fixpoint();
+
+        results.visit_reachable_with(
+            mir,
+            &mut NeedlessCowVisitor {
+                cx,
+                cow_locals_to_assignments: &mut cow_locals_to_assignments,
+            },
+        );
+
+        for assignments in cow_locals_to_assignments.values() {
+            if assignments.len() == 1 {
+                continue;
+            }
+            let Some(first) = assignments.first() else {
+                continue;
+            };
+
+            span_lint_hir_and_then(
+                cx,
+                NEEDLESS_COW,
+                mir.source_scopes[first.source_info().scope]
+                    .local_data
+                    .as_ref()
+                    .assert_crate_local()
```

Comment:
This call feels suspicious. Isn't `ClearCrossCrate` meant to prevent usage from other crates (In this case, I suppose clippy)? For `local_info` this would ICE. Was this not cleared so that clippy can access this?